Code/GraphMol/ScaffoldNetwork/ScaffoldNetwork.cpp

```diff
@@ -0,0 +1,437 @@
+//
+//  Copyright (C) 2019 Greg Landrum and T5 Informatics GmbH
+//
+//   @@ All Rights Reserved @@
+//  This file is part of the RDKit.
+//  The contents are covered by the terms of the BSD license
+//  which is included in the file license.txt, found at the root
+//  of the RDKit source tree.
+//
+#include "ScaffoldNetwork.h"
+#include "detail.h"
+
+#include <boost/shared_ptr.hpp>
+#include <RDGeneral/types.h>
+#include <RDGeneral/Invariant.h>
+#include <GraphMol/RDKitBase.h>
+#include <GraphMol/ROMol.h>
+#include <GraphMol/MolOps.h>
+#include <GraphMol/ChemReactions/Reaction.h>
+#include <GraphMol/ChemReactions/ReactionParser.h>
+#include <GraphMol/SmilesParse/SmilesWrite.h>
+#include <GraphMol/MolStandardize/Fragment.h>
+#include <GraphMol/ChemTransforms/ChemTransforms.h>
+#include <RDGeneral/Exceptions.h>
+
+namespace RDKit {
+namespace ScaffoldNetwork {
+
+ScaffoldNetworkParams::ScaffoldNetworkParams(
+    const std::vector<std::string> &bondBreakersSmarts) {
+  bondBreakersRxns.clear();
+  bondBreakersRxns.reserve(bondBreakersSmarts.size());
+  for (auto sma : bondBreakersSmarts) {
+    std::shared_ptr<ChemicalReaction> rxn(RxnSmartsToChemicalReaction(sma));
+    if (rxn.get() == nullptr) {
+      throw ValueErrorException("could not parse reaction smarts: " + sma);
+    }
+    if (rxn->getNumReactantTemplates() != 1) {
+      throw ValueErrorException(
+          "bond breaker reactions must have exactly one reactant");
+    }
+    if (rxn->getNumProductTemplates() != 2) {
+      throw ValueErrorException(
+          "bond breaker reactions must have exactly two products");
+    }
+    rxn->initReactantMatchers();
+    bondBreakersRxns.emplace_back(rxn);
+  }
+};
+
+namespace detail {
+
+void updateMolProps(RWMol &mol, const ScaffoldNetworkParams &params) {
+  RDUNUSED_PARAM(params);
+  MolOps::sanitizeMol(mol);
+}
+
+ROMol *makeScaffoldGeneric(const ROMol &mol, bool doAtoms, bool doBonds) {
```

Comment:
Perhaps this function could be made into a functor for easier replacement, i.e. if we wanted to use one of rogers other hashes for instance.

Reply:
I like the idea of switching over to using functors to control the whole process (basically describing a workflow for the fragmentation)
I can see that being super useful, but also a non-trivial amount of work (and testing). Ok to get this basic version (which is already useful) in first and then think about that as a future project?